drive/src/drive/flags.rs

```diff
@@ -30,44 +30,620 @@
 //! Flags
 //!
 
-use byteorder::{BigEndian, ReadBytesExt};
+use crate::drive::flags::StorageFlags::{
+    MultiEpoch, MultiEpochOwned, SingleEpoch, SingleEpochOwned,
+};
+use costs::storage_cost::removal::StorageRemovedBytes::SectionedStorageRemoval;
+use costs::storage_cost::removal::{StorageRemovalPerEpochByIdentifier, StorageRemovedBytes};
 use grovedb::ElementFlags;
+use integer_encoding::VarInt;
+use intmap::IntMap;
+use std::cmp::Ordering;
+use std::collections::BTreeMap;
 
-use crate::error::drive::DriveError;
+use crate::error::storage_flags::StorageFlagsError;
 use crate::error::Error;
 
+type EpochIndex = u16;
+
+type BaseEpoch = EpochIndex;
+
+type BytesAddedInEpoch = u32;
+
+type OwnerId = [u8; 32];
+
 /// Storage flags
-#[derive(Clone, Debug, Default, PartialEq)]
-pub struct StorageFlags {
-    /// Epoch
-    pub epoch: u16,
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum StorageFlags {
+    /// Single epoch
+    /// represented as byte 0
+    SingleEpoch(BaseEpoch),
+
+    /// Multi epoch
+    /// represented as byte 1
+    MultiEpoch(BaseEpoch, BTreeMap<EpochIndex, BytesAddedInEpoch>),
+
+    /// Single epoch owned
+    /// represented as byte 2
+    SingleEpochOwned(BaseEpoch, OwnerId),
+
+    /// Multi epoch owned
+    /// represented as byte 3
+    MultiEpochOwned(BaseEpoch, BTreeMap<EpochIndex, BytesAddedInEpoch>, OwnerId),
 }
 
 impl StorageFlags {
+    /// Create new single epoch storage flags
```

Comment:
Use state pattern it's gonna make StorageFlags cleaner (less nesting in functions) and smaller:
- https://refactoring.guru/design-patterns/state/rust/example
- https://doc.rust-lang.org/book/ch17-03-oo-design-patterns.html

We could add a todo if you agree and refactor later to do not delay v23

Reply:
The state pattern describes events that change the state of the structure. This isn't applicable to StorageFlags as their state isn't changed from a simple event but is changed through a user based state transition.

You can see the state pattern as being A - ST -> B. An example that was given was that of a music player. It can be in many states by creating simple state transitions. Like Play and Stop. This isn't applicable to StorageFlags. They don't change states from a simple event. Instead they are used for storage for complex state and fee information.

Reply:
In the book another example. I think it's not important who change the state, the important part is that you can encapsulate logic into specific structures that represents different states so `StorageFlags` become much simpler. Also you will be able easily unit test each state separately and then test `StorageFlags` with mocks.

I'm not insisting to do that now, because it's a lot of work but we could add a todo and do it later.

Reply:
You are free to propose a PR in the future as I can't really conceptualize what you are seeing. I don't think it's a good idea to add this to the backlog because I do not believe it would be easy for someone else to see it. I personally do not think it's applicable here but could be proven wrong.

Reply:
Instead of Trait better use Enum is it will validated in compile time

Reply:
Ok, we will come back to it.